mosaic-runtime/src/commonMain/kotlin/com/jakewharton/mosaic/mosaic.kt

```diff
@@ -155,21 +158,82 @@ internal class MosaicComposition(
 	coroutineScope: CoroutineScope,
 	private val terminalState: State<Terminal>,
 	private val keyEvents: ReceiveChannel<KeyEvent>,
-	onEndChanges: (MosaicNode) -> Unit,
+	private val onDrawing: (MosaicNode) -> Unit,
 ) {
 	private val job = Job(coroutineScope.coroutineContext[Job])
 	private val clock = BroadcastFrameClock()
 	private val composeContext: CoroutineContext = coroutineScope.coroutineContext + job + clock
 	val scope = CoroutineScope(composeContext)
 
-	private val applier = MosaicNodeApplier(onEndChanges)
+	private val applier = MosaicNodeApplier(::doLayout)
 	val rootNode = applier.root
 	private val recomposer = Recomposer(composeContext)
 	private val composition = Composition(applier, recomposer)
 
+	private val applyObserverHandle: ObserverHandle
+
+	private val readingStatesOnLayout = mutableSetOf<Any>()
+	private val readingStatesOnDrawing = mutableSetOf<Any>()
```

Comment:
We might want to use the new collections inside androidx.collection to get more efficiency from this in the future. Can be a follow-up (and I'll add it to #190).